src/ExactDiagonalization/exact_diagonalization_problem.jl

```diff
@@ -146,4 +205,65 @@ function Base.:(==)(p1::ExactDiagonalizationProblem, p2::ExactDiagonalizationPro
         p1.kwargs == p2.kwargs
 end
 
-Rimu.Hamiltonians.dimension(p::ExactDiagonalizationProblem) = dimension(p.hamiltonian)
+Rimu.Hamiltonians.dimension(p::ExactDiagonalizationProblem) = p.linear_dimension
+
+function _set_up_starting_address(v0, ham)
+    if isnothing(v0)
+        addr_or_vec = starting_address(ham)
+    elseif allows_address_type(ham, v0) ||
+           v0 isa Union{NTuple,Vector} && allows_address_type(ham, eltype(v0))
+        addr_or_vec = v0
+    elseif v0 isa FrozenDVec
+        addr_or_vec = keys(v0)
+    else
+        throw(ArgumentError("Invalid starting vector in `ExactDiagonalizationProblem`."))
+    end
+
+    return addr_or_vec # single address or iterable of addresses
+end
+
+"""
+    estimate_memory_requirement(::ExactDiagonalizationProblem)
+    estimate_memory_requirement(h::AbstractHamiltonian, linear_dimension=dimension(h))
+    -> (; dense, matrix_free, sparse)
+
+Estimate the memory requirement for an [`ExactDiagonalizationProblem`](@ref).
+This function estimates the memory requirement based on the linear dimension and the
+Hamiltonian. It returns an estimate of the memory size in bytes for three
+different types of algorithms:
+- `dense`: The memory requirement for [`LinearAlgebraSolver()`](@ref) using a dense matrix.
+- `matrix_free`: The memory requirement for [`KrylovKitSolver(; matrix_free=true)`](@ref)
+  using a matrix-free algorithm.
+- `sparse`: The memory requirement for [`KrylovKitSolver(; matrix_free=false)`](@ref)
+  using a sparse matrix.
+
+The memory requirements for other sparse solvers are expected to be similar to the
+`KrylovKitSolver` estimates.
+"""
+function estimate_memory_requirement(prob::ExactDiagonalizationProblem)
+    return estimate_memory_requirement(prob.hamiltonian, prob.linear_dimension)
+end
+
+function estimate_memory_requirement(
+    hamiltonian::AbstractHamiltonian,
+    linear_dimension=dimension(hamiltonian)
+)
+    krylovdim = 30
+    # standard Krylov dimension in KrylovKit, could be read from the algorithm
```

Comment:
I agree it should be algorithm dependent. Since KrylovKit allocates new vectors on every allocation, I'm not sure how to accurately estimate the amount of memory required there. Arpack might be operating in-place, but I'm not sure.

Reply:
In particular LOBPCG should be very memory efficient. It should only need two or three instances of the vector at a time.